File: src/main/java/dev/psygamer/wireframe/util/collection/FreezableHashSet.java
package dev.psygamer.wireframe.util.collection;

import java.util.*;

public class FreezableHashSet <E> extends FreezableSet<E> {
	
	public FreezableHashSet() {
		super(new HashSet<>());
	}
	
	public FreezableHashSet(final int initialCapacity) {
		super(new HashSet<>(initialCapacity));
	}
	
	public FreezableHashSet(final int initialCapacity, final float loadFactor) {
		super(new HashSet<>(initialCapacity, loadFactor));
	}
	
	public FreezableHashSet(final Collection<? extends E> set) {
		super(new HashSet<>(set));
	}
	
	@Override
	public FreezableSet<E> copy() {
		final FreezableSet<E> set = new FreezableHashSet<>(this);
		
		if (isFrozen())
			set.freeze();
		
		return set;
	}
}

File: src/main/kotlin/dev/psygamer/wireframe/util/collection/FreezableHashSet.kt
package dev.psygamer.wireframe.util.collection

class FreezableHashSet<E> : FreezableSet<E> {

	constructor() : super(HashSet<E>())
	constructor(initialCapacity: Int) : super(HashSet<E>(initialCapacity))
	constructor(initialCapacity: Int, loadFactor: Float) : super(HashSet<E>(initialCapacity, loadFactor))
	constructor(set: Set<E>) : super(HashSet<E>(set))
}